lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+// Owns Tasks, handles correspondence with FDs and pids.
+// FIXME: only handles stdout fds.
+class TaskMap {
+  using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
+
+public:
+  TaskMap() = default;
+
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
+
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
+
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
+
+  void eraseTask(Task &T) { TasksByPid.erase(T.getPid()); }
+};
+
+/// Concurrently execute the tasks in the TaskQueue, collecting the outputs from
+/// each task. The typical task is a Swift frontend job.
+/// Maintain invarients connecting tasks to execute, tasks currently executing,
+/// and fds being polled. Handle output from and death of tasks.
+class TaskShepherd {
+
+  /// The set of tasks needed to be executed.
+  std::queue<std::unique_ptr<Task>> &TasksToExecute;
+
+public:
+  /// Unix system calls deal in process IDs (Pids), so a means is needed
```

Comment:
"Pid" really feels like a non-standard capitalization here (cf. [`getpid()`'s man page](http://man7.org/linux/man-pages/man2/getpid.2.html)). It didn't bother me so much in variable names but it looks really bad to me in prose.

Reply:
Nice! Prose "Pid" and "pid" replaced by "process ID".